snuba/clusters/cluster.py

```diff
@@ -61,16 +72,29 @@ class ClickhouseCluster(Cluster[SqlQuery, ClickhouseWriterOptions]):
     shared by all storages located on the cluster
     """
 
-    def __init__(self, host: str, port: int, http_port: int, storage_sets: Set[str]):
+    def __init__(
+        self,
+        host: str,
+        port: int,
+        http_port: int,
+        storage_sets: Set[str],
+        is_distributed: bool,
```

Comment:
I am not sure the `is_distributed` name is appropriate. We may have replication only, distributed may imply that we distribute data across nodes which means sharding.
`single_node` may be a better name to convey this meaning.

Taking a step back, I am not sure I am fully confident about which specific cases we need concept of distributed mode for. I remember only this:
- table name (which would be different)

Am I correct ?

Reply:
Right now it's only table name. In future, the distributed (or non single_node) option will also be used to determine the set of storage nodes that would be used for any of the commands that will need to also operate on them (cleanup/optimize/migrate).

Reply:
"will also be used to determine the set of storage nodes that would be used for any of the commands that will need to also operate on them"

I don't think I have clear how this will work. If you have a set of storage nodes and a set of query nodes in your cluster why do you need this flag to determine the set of storage nodes you would run those commands?

Reply:
What do you mean by "set of query nodes in your cluster"? As far as Snuba application is concerned I think there is only one query node per cluster (this is going to be the node with the distributed table in the distributed mode). Right now we only have one global query node, but we are moving towards one per cluster.

The idea here is that the user would only specify the host and port of the query node for the cluster. The storage nodes are discovered by going to the `system.clusters` table on the query node if it's a distributed cluster. Otherwise the storage node is just the query node if it's not distributed.

Reply:
I think there can be multiple query nodes in a distributed environment for a given clusters, on different hosts. Not all Clickhouse queries need to go through the same physical host. These nodes strictly speaking are not part of the clickhosue cluster (they do not show up in clickhouse configuration) in clickhouse terminology. 
Were you planning to expose the fact that there may be physically be multiple query nodes or to hide this in snuba ?

Reply:
After giving this some more thought, I think I've changed my mind about the need to distinguish storage and query nodes - I don't think a Snuba cluster needs to know about more than one node which is used for both storage and querying at all.

For a SELECT query, Snuba only ever connects to the one node in Clickhouse (reader always comes this). It's Clickhouse's job to either just run that (if it's local table) or figure out which of the one or combination of remote servers should actually run the query based on  sharding key, settings, Clickhouse configuration, etc (if it's distributed table).

INSERTs are the same. Again Snuba only knows about the one node with the distributed table. It's Clickhouse's job to figure out which actual servers need to get updated with the inserted data.

So the `get_nodes()` method introduced here is neither for storage nor query. It's for performing specific operations (like migrations) where we must actually connect to all of the individual hosts, so we need to discover what they are.

Does this make sense to you? I know it's different from what we discussed earlier about storage nodes vs query nodes, but I can't really remember the reason why that distinction was needed anymore.

Reply:
It seems to me your proposal is not really different from the original design. If a cluster contains a set of nodes used for DDL operations like migrations and still define one node used for insert and query, isn't this last node the query node? It still needs to be identified separately in the cluster configuration or in the snuba configuration so a distinction still needs to happen.

There is a secondary problem though (in both this proposal and the original design). How do we deal with multiple query nodes? We can do that to distribute the load. Are you planning to support this at all ? Or is snuba going to just know about one node to send query to ?
If we expose the existence of multiple nodes snuba will need to take care of load balancing (which it does not today). If we expose only one we need to give directions on how to do load balancing.

Reply:
Yeah, I'm suggesting something close to what we have now.

Are you proposing moving load balancing responsibilities into Snuba as part of this? That is not even something I'd considered at all (or really want to unless there's a compelling reason). What do you think the benefits of that would be would be? I was under the impression Clickhouse generally does a good job of this.

I think the only directions we give today are "in_order" or "random", is there some other directions you think would be required here even if we stick with the exposing single node approach?

Reply:
oh no, I don't think we should conflate clickhouse load balancing here, but we need to decide which directions to give. We could make snuba only be aware of one query node, this node can be a real host or a proxy that redirects to multiple clickhosue distributed table nodes so snuba does not need to know the existence of multiple clickhouse nodes with the distributed table deployed.
Though we need to be deliberate here.

The load balancer I am talking about is the one between snuba and clickhosue query nodes, not the one between clickhosue query nodes and clickhouse storage nodes.

Reply:
I added some documentation about this to the class. Let me know if it can also be clearer through the design.

Reply:
This approach should also be shared with the team in the design document, or with an amended document (probably there is no need for a meeting).

Reply:
have you verified how we would run the distributed table DDL operations? If snuba does not see all clickhouse nodes and they are all hidden behind one address snuba cannot be certain it created the distributed table on all nodes.

Reply:
I think there's 2 options to find those nodes.
1. Tell each Snuba cluster about those hosts via settings
2. Add that mapping to the clickhouse config and tell each Snuba cluster about that Clickhouse "cluster_name" so it can use that to look up those hosts

Reply:
`Add that mapping to the clickhouse config ` I don't think those nodes are in the clickhouse config. I think the config option is better now the questions are:
1) are these nodes part of the cluster data structure? If not how are they represented in terms of data structures? How does a migration script deal with them when running a DDL operation?
2) does the snuba cluster need to know them or is it just for the script running DDL operations.

Reply:
Yes, they'll need to be part of the cluster (and accessed by the script running the DDL operation). Even though it's not in the config yet, i'll add some code here showing how I think it can work.